### PR TITLE
Query Frontend: new service for queuing and retrying queries.

### PR DESCRIPTION
**Pull request by tomwilkie**
A new job which accepts HTTP requests and queues them by tenant ID, retrying them on errors.  Aim is to allow for the occasional large query to cause a querier OOM without failing requests, allowing us to over-provision querier parallelism.  Also, prevent multiple large request from blocking a single querier by distributing them FIFO across all queriers.  And finally, prevent a single tenant from DoSing other tenants by fairly scheduling queries between tenants.

The query frontend job accepts gRPC streaming requests from the queriers, which then "pull" requests from the frontend.  For HA it is recommended you run multiple frontends - the queriers will connect to (and pull requests from) both.  To get the benefit of the fair scheduling, it is recommended you run fewer frontends than queriers - two should suffice.

See https://docs.google.com/document/d/1lsvSkv0tiAMPQv-V8vI2LZ8f4i9JuTRsuPI_i-XcAqY for more details design discussion.  In the future, query splitting, query alignment and query results caching will be added to the frontend.

Signed-off-by: Tom Wilkie <tom.wilkie@gmail.com>

### CODE REVIEW

File: pkg/querier/frontend/frontend.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+		Help:      "Number of times a request is retried.",
+		Buckets:   []float64{0, 1, 2, 3, 4, 5},
+	})
+	queueLength = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "cortex",
+		Name:      "query_frontend_queue_length",
+		Help:      "Number of queries in the queue.",
+	})
+
+	errServerClosing  = httpgrpc.Errorf(http.StatusTeapot, "server closing down")
```
Comment (csmarchbanks) [inline pkg/querier/frontend/frontend.go]: Nice status code

**Comment by csmarchbanks** [comment]
Overall this looks good to me. Would you be willing to add a doc that describes the two options between running just a querier, and running a front-end + queriers are? Mostly what is in your PR description, but somewhere that is easier to find in the future.

File: pkg/querier/frontend/frontend.go (hunk 1 above)
Comment (tomwilkie) [reply pkg/querier/frontend/frontend.go]: I was hoping for a official "closing down" status code, but I couldn't find anything...

**Comment by tomwilkie** [comment]
@csmarchbanks PTAL, added some documentation.

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+- KubeCon Prometheus Day talk "Weave Cortex: Multi-tenant, horizontally scalable Prometheus as a Service" [slides](http://www.slideshare.net/weaveworks/weave-cortex-multitenant-horizontally-scalable-prometheus-as-a-service) [video](https://www.youtube.com/watch?v=9Uctgnazfwk)
+- PromCon 2017 Talk: "[Cortex: Prometheus as a Service, One Year On](https://promcon.io/2017-munich/talks/cortex-prometheus-as-a-service-one-year-on/)"
+- CNCF TOC Presentation; "Horizontally Scalable, Multi-tenant Prometheus" [slides](https://docs.google.com/presentation/d/190oIFgujktVYxWZLhLYN4q8p9dtQYoe4sxHgn4deBSI/edit#slide=id.g3b8e2d6f7e_0_6)
+
+## Query Path
+
+### Query Frontend
+
+The query frontend is an optional job which accepts HTTP requests and queues them by tenant ID, retrying them on errors. This allow for the occasional large query which would otherwise cause a querier OOM, allowing us to over-provision querier parallelism. Also, it prevents multiple large from being convoyed on a single querier by distributing them FIFO across all queriers. And finally, it prevent a single tenant from DoSing other tenants by fairly scheduling queries between tenants.
```
Comment (csmarchbanks) [inline docs/architecture.md]: You are missing a word after "prevents multiple large"

**Comment by tomwilkie** [comment]
@csmarchbanks thanks, PTAL

File: vendor/github.com/weaveworks/common/httpgrpc/server/server.go [hunk 3]
```diff
@@ -66,7 +66,7 @@ func (s Server) Handle(ctx context.Context, r *httpgrpc.HTTPRequest) (*httpgrpc.
 	if recorder.Code/100 == 5 {
 		return nil, httpgrpc.ErrorFromHTTPResponse(resp)
 	}
-	return resp, err
+	return resp, nil
```
Comment (bboreham) [inline vendor/github.com/weaveworks/common/httpgrpc/server/server.go]: Nothing should be changed in `vendor` without going via `dep`.

**Pull request merged**